File: output/windows/cooked/code/gameplay/DirectionalShadowSunSweepComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "DirectionalShadowSunSweepComponent.hpp"
#include "UpdateComponent.hpp"
#include "float4.hpp"
#include "system/math.hpp"

DirectionalShadowSunSweepComponent::DirectionalShadowSunSweepComponent() : MaxYawRadians(), MinYawRadians(), PitchRadians(), SweepSpeedRadians()
{
}

float DirectionalShadowSunSweepComponent::get_MaxYawRadians()
{
return this->MaxYawRadians;
}

void DirectionalShadowSunSweepComponent::set_MaxYawRadians(float value)
{
this->MaxYawRadians = value;
}

float DirectionalShadowSunSweepComponent::get_MinYawRadians()
{
return this->MinYawRadians;
}

void DirectionalShadowSunSweepComponent::set_MinYawRadians(float value)
{
this->MinYawRadians = value;
}

float DirectionalShadowSunSweepComponent::get_PitchRadians()
{
return this->PitchRadians;
}

void DirectionalShadowSunSweepComponent::set_PitchRadians(float value)
{
this->PitchRadians = value;
}

float DirectionalShadowSunSweepComponent::get_SweepSpeedRadians()
{
return this->SweepSpeedRadians;
}

void DirectionalShadowSunSweepComponent::set_SweepSpeedRadians(float value)
{
this->SweepSpeedRadians = value;
}

void DirectionalShadowSunSweepComponent::Update()
{
UpdateComponent::Update();
const double normalized = (Math::Sin(Core->Instance->TotalElapsedSeconds * this->SweepSpeedRadians) * 0.5) + 0.5;
const double yawRadians = this->MinYawRadians + ((this->MaxYawRadians - this->MinYawRadians) * normalized);
float4 orientation;
float4::CreateFromYawPitchRoll(static_cast<float>(yawRadians), this->PitchRadians, 0.0f, orientation);
orientation.Normalize();
Parent->LocalOrientation = orientation;
}

File: assets/codebase/rendering/DirectionalShadowSunSweepComponent.cs
namespace gameplay.rendering {
    /// <summary>
    /// Sweeps one directional light through a narrow sun arc using a sine wave over absolute runtime time.
    /// </summary>
    public sealed class DirectionalShadowSunSweepComponent : UpdateComponent {
        /// <summary>
        /// Stores the accumulated local elapsed time used to evaluate the authored sweep phase.
        /// </summary>
        double ElapsedSeconds;

        /// <summary>
        /// Gets or sets the minimum yaw reached by the light sweep in radians.
        /// </summary>
        public float MinYawRadians { get; set; }

        /// <summary>
        /// Gets or sets the maximum yaw reached by the light sweep in radians.
        /// </summary>
        public float MaxYawRadians { get; set; }

        /// <summary>
        /// Gets or sets the fixed pitch applied throughout the sweep in radians.
        /// </summary>
        public float PitchRadians { get; set; }

        /// <summary>
        /// Gets or sets the angular sweep rate in radians per second.
        /// </summary>
        public float SweepSpeedRadians { get; set; }

        /// <summary>
        /// Resets the local sweep timer when the component joins a scene entity.
        /// </summary>
        /// <param name="entity">Owning entity receiving the component.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            ElapsedSeconds = 0d;
        }

        /// <summary>
        /// Evaluates the current light orientation from accumulated local runtime time.
        /// </summary>
        public override void Update() {
            base.Update();

            ElapsedSeconds += Core.Instance.FrameDeltaSeconds;
            double normalized = (Math.Sin(ElapsedSeconds * SweepSpeedRadians) * 0.5d) + 0.5d;
            double yawRadians = MinYawRadians + ((MaxYawRadians - MinYawRadians) * normalized);
            float4 orientation;
            float4.CreateFromYawPitchRoll((float)yawRadians, PitchRadians, 0f, out orientation);
            orientation.Normalize();
            Parent.LocalOrientation = orientation;
        }
    }
}
